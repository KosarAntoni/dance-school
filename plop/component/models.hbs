export type {{properCase name}}Props {

}