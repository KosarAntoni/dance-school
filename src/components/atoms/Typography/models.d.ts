type Breakpoints = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl';

type LineHeight = 'small' | 'normal' | 'big';

type FontWeight = 'light' | 'bold' | 'black' | string;

type Component = 'h1' | 'h2' | 'h3' | 'p' | 'span' | 'div';

type Size =
  | 8
  | 10
  | 12
  | 14
  | 15
  | 16
  | 17
  | 18
  | 20
  | 24
  | 25
  | 28
  | 30
  | 35
  | 40
  | 42
  | 50
  | 55
  | 65
  | number;
interface SizeBreakpoints extends Partial<Record<Breakpoints, Size>> {
  base: Size;
}

type Position = 'top' | 'right' | 'bottom' | 'left';
type Padding = 'xxs' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl' | 'xxxl' | 'huge';
interface PaddingPosition extends Partial<Record<Position, Padding>> {}
interface PaddingBreakpoints extends Partial<Record<Breakpoints, Padding>> {
  base: Padding;
}
interface PaddingBreakpointsPosition extends Partial<Record<Position, PaddingBreakpoints>> {}

type Align = 'inherit' | 'left' | 'center' | 'right';
interface AlignBreakpoints extends Partial<Record<Breakpoints, Align>> {
  base: Align;
}

export interface TypographyProps {
  size?: Size | SizeBreakpoints;
  padding?: Padding | PaddingPosition | PaddingBreakpointsPosition;
  align?: Align | AlignBreakpoints;
  as?: Component;
  className?: string;
  lineHeight?: LineHeight;
  weight?: FontWeight;
  italic?: boolean;
  dangerouslySetInnerHTML?: string;
}
