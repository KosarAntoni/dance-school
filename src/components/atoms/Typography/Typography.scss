@import 'styles/component-set';
@import './utils';

.typography {
  font-family: $font-base;
  font-weight: 400;
  position: relative;
  display: inline-block;

  &--light {
    font-weight: 300;
  }

  &--bold {
    font-weight: 700;
  }

  &--italic {
    font-style: italic;
  }

  &--heading {
    font-family: $font-heading;
    font-weight: 700;
  }
}

@include line-height;

@include align;

@each $breakpoint in map-keys($breakpoints) {
  @include align($breakpoint);
}

@include padding;

@each $breakpoint in map-keys($breakpoints) {
  @include padding($breakpoint);
}

@include size;

@each $breakpoint in map-keys($breakpoints) {
  @include size($breakpoint);
}
