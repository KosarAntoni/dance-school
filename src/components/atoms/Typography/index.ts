export * from './models.d';
export { default } from './Typography';
