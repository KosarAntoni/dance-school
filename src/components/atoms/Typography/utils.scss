// Variables

$line-heights-map: (
  big: 1.3,
  normal: 1.1,
  small: 0.9,
);
$paddings-map: (
  unset: unset,
  xxs: $gap-xxsmall,
  xs: $gap-xsmall,
  sm: $gap-small,
  md: $gap-regular,
  lg: $gap-large,
  xl: $gap-xlarge,
  xxl: $gap-xxlarge,
  xxxl: $gap-xxxlarge,
  huge: $gap-huge,
);
$font-sizes-list: 8, 10, 12, 14, 15, 16, 17, 18, 20, 24, 25, 28, 30, 35, 40, 42, 50, 55, 65;
$sides-list: 'top', 'left', 'bottom', 'right';
$aligns-list: 'inherit', 'left', 'center', 'right';

// Mixins

@mixin line-height() {
  @each $line-height, $value in $line-heights-map {
    .typography--line-height-#{$line-height} {
      line-height: $value;
    }
  }
}

@mixin padding($breakpoint: 'none') {
  @if $breakpoint != 'none' {
    @each $size, $value in $paddings-map {
      .typography--padding-#{$breakpoint}-#{$size} {
        @include breakpoint(#{$breakpoint}-up) {
          padding: $value;
        }
      }
    }
    @each $side in $sides-list {
      @each $size, $value in $paddings-map {
        .typography--padding-#{$side}-#{$breakpoint}-#{$size} {
          @include breakpoint(#{$breakpoint}-up) {
            padding-#{$side}: $value;
          }
        }
      }
    }
  } @else {
    @each $size, $value in $paddings-map {
      .typography--padding-#{$size} {
        padding: $value;
      }
    }
    @each $side in $sides-list {
      @each $size, $value in $paddings-map {
        .typography--padding-#{$side}-#{$size} {
          padding-#{$side}: $value;
        }
      }
    }
  }
}

@mixin align($breakpoint: 'none') {
  @each $align in $aligns-list {
    @if $breakpoint != 'none' {
      .typography--align-#{$breakpoint}-#{$align} {
        @include breakpoint(#{$breakpoint}-up) {
          text-align: #{$align};
        }
      }
    } @else {
      .typography--align-#{$align} {
        text-align: #{$align};
      }
    }
  }
}

@mixin size($breakpoint: 'none') {
  @each $size in $font-sizes-list {
    @if $breakpoint != 'none' {
      .typography--size-#{$breakpoint}-#{$size} {
        @include breakpoint(#{$breakpoint}-up) {
          @include font-size($size);
        }
      }
    } @else {
      .typography--size-#{$size} {
        @include font-size($size);
      }
    }
  }
}
