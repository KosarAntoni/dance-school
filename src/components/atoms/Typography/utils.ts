import isNumber from 'lodash/isNumber';
import isPlainObject from 'lodash/isPlainObject';
import isString from 'lodash/isString';

import {
  Align,
  AlignBreakpoints,
  Padding,
  PaddingBreakpointsPosition,
  PaddingPosition,
  Size,
  SizeBreakpoints,
} from './models.d';

export const sizeClassNames = (size: Size | SizeBreakpoints): string[] => {
  if (isNumber(size)) {
    return [`typography--size-${size}`];
  }
  if (isPlainObject(size) && size.base) {
    return Object.entries(size).map(([breakpoint, value]): string =>
      breakpoint === 'base'
        ? `typography--size-${value}`
        : `typography--size-${breakpoint}-${value}`
    );
  }

  return [];
};

export const alignClassNames = (align: Align | AlignBreakpoints): string[] => {
  if (isString(align)) {
    return [`typography--align-${align}`];
  }
  if (isPlainObject(align) && align.base) {
    return Object.entries(align).map(([breakpoint, value]): string =>
      breakpoint === 'base'
        ? `typography--align-${value}`
        : `typography--align-${breakpoint}-${value}`
    );
  }

  return [];
};

export const paddingClassNames = (
  padding: Padding | PaddingPosition | PaddingBreakpointsPosition
): string[] => {
  const paddingPositions: string[] = ['top', 'right', 'bottom', 'left'];

  if (isString(padding)) {
    return paddingPositions.map((position): string => `typography--padding-${position}-${padding}`);
  }
  if (isPlainObject(padding)) {
    return Object.entries(padding)
      .map(([side, value]) => {
        if (isString(value)) {
          return `typography--padding-${side}-${value}`;
        }
        if (isPlainObject(value) && value.base) {
          return Object.entries(value).map(([breakpoint, paddingValue]) =>
            breakpoint === 'base'
              ? `typography--padding-${side}-${paddingValue}`
              : `typography--padding-${side}-${breakpoint}-${paddingValue}`
          );
        }

        return '';
      })
      .flat();
  }

  return [];
};
