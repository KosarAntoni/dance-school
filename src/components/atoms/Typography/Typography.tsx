import React, { FC } from 'react';
import classNames from 'classnames';

import { alignClassNames, paddingClassNames, sizeClassNames } from './utils';

import { TypographyProps } from './models.d';

import './Typography.scss';

/**
 * @example <Typography as = 'p'
 *                      className = 'testClass'
 *                      align = {{ base: 'center', sm: 'left', md: 'right', xl: 'inherit' }}
 *                      padding = {{ top: 'sm', right: 'md', bottom: 'lg', left: 'xl' }}
 *                      lineHeight = 'small'
 *                      weight = 'light'
 *                      italic
 *                      size = {{ base: 8, xs: 10, sm: 12, md: 14, lg: 16, xl: 18, xxl: 20 }} >
 *  Awesome text example
 * </Typography>
 */
const Typography: FC<TypographyProps> = ({
  as /** 'h1' | 'h2' | 'h3' | 'p' | 'span' | 'div */,
  align /** 'inherit' | 'left' | 'center' | 'right' | { @requires base: value, @optional breakpoint: value} */,
  lineHeight = 'normal' /** 'small' | 'normal' | 'big' */,
  weight /** 'light' | 'bold' | 'black' */,
  italic /** boolean */,
  size /** 8 | 10 | 12 | 14 | 15 | 16 | 17 | 18 | 20 | 24 | 25 | 28 | 30 | 35 | 40 | 42 | 50 | 55 | 65 | { @requires base: value, @optional breakpoint: value} */,
  padding /** based on varables placed in "styles/common/variables" 'xxs' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl' | 'xxxl' | 'huge' | { top: value, right: value, bottom: value, left: value } | { side: { @requires base: value, @optional breakpoint: value} } */,
  className /** string */,
  children,
  ...rest
}) => {
  const Component = as || 'span';

  const sizeClasses = size ? sizeClassNames(size) : [];
  const paddingClasses = padding ? paddingClassNames(padding) : [];
  const alignClasses = align ? alignClassNames(align) : [];

  const componentClasses = classNames(
    'typography',
    className,
    ...sizeClasses,
    ...paddingClasses,
    ...alignClasses,
    {
      [`typography--line-height-${lineHeight}`]: lineHeight,
      [`typography--${weight}`]: weight,
      'typography--italic': italic,
    }
  );

  return (
    <Component data-testid="typography-item" className={componentClasses} {...rest}>
      {children}
    </Component>
  );
};

export default Typography;
