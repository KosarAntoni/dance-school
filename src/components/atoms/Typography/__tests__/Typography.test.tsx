import React from 'react';
import { render, screen } from '@testing-library/react';

import {
  TypographyChildrenMock,
  typographyDefaultClassesMock,
  TypographyDefaultPropsMock,
  typographyExtendedClassesMock,
  TypographyExtendedPropsMock,
} from '../__mock__/mock';
import Typography from '../Typography';

import { TypographyProps } from '../models.d';

const defaultProps: TypographyProps = {
  ...(TypographyDefaultPropsMock as TypographyProps),
};

const extendedProps: TypographyProps = {
  ...(TypographyExtendedPropsMock as TypographyProps),
};

it('Typography component renders correctly with required and optional props', () => {
  const { rerender } = render(<Typography {...defaultProps}>{TypographyChildrenMock}</Typography>);

  let typographyNode = screen.getByRole('heading');

  // the component has been rendered
  expect(typographyNode).toBeInTheDocument();

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(typographyNode).toMatchSnapshot();

  // the component renders all classes
  expect(typographyNode).toHaveClass(typographyDefaultClassesMock.join(' '));

  rerender(<Typography {...extendedProps}>{TypographyChildrenMock}</Typography>);

  typographyNode = screen.getByTestId('typography-item');

  // the component renders all classes
  expect(typographyNode).toHaveClass(typographyExtendedClassesMock.join(' '));

  // it should not be present as heading when 'as' prop is provided as paragraph
  expect(screen.queryByRole('heading')).toBeNull();

  rerender(<Typography>{TypographyChildrenMock}</Typography>);

  typographyNode = screen.getByTestId('typography-item');

  // the component renders without classes
  expect(typographyNode).not.toHaveClass(typographyDefaultClassesMock.join(' '));
  expect(typographyNode).not.toHaveClass(typographyExtendedClassesMock.join(' '));
});
