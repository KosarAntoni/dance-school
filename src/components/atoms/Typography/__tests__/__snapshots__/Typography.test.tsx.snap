// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography component renders correctly with required and optional props 1`] = `
<h1
  class="typography testClass typography--size-8 typography--padding-top-xl typography--padding-right-xl typography--padding-bottom-xl typography--padding-left-xl typography--align-center typography--line-height-big typography--bold typography--italic"
  data-testid="typography-item"
>
  Test Lorem Ipsum
</h1>
`;
